File: tsm-rossler.c
/*
 * Rossler System
 *
 * (c) 2018-2025 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
 */
#include <stdio.h>
#include <stdlib.h>
#include <mpfr.h>
#include "taylor-ode.h"

struct Parameters { real a, b, c; };

model *tsm_init_p (int argc, char **argv, int n) { (void)n;
    CHECK(argc == 12);
    model *_ = malloc(sizeof (model)); CHECK(_);
    tsm_get_p(argv, argc, &_->a, &_->b, &_->c);
    return _;
}

void ode (triplet *v, series x, series y, series z, const model *_, const int k) {
    //  x' = - y - z
    mpfr_add(v->x, y[k], z[k], RND);
    mpfr_neg(v->x, v->x, RND);
    //  y' = x + Ay
    mpfr_fma(v->y, _->a, y[k], x[k], RND);
    //  z' = B + z(x - C)
    mpfr_fms(v->z, _->c, z[k], *t_mul(z, x, k), RND);
    mpfr_neg(v->z, v->z, RND);
    if (!k) mpfr_add(v->z, v->z, _->b, RND);
}

File: tsm-rossler.py
#!/usr/bin/env python3
#
#  (c) 2018-2023 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
import print_args
from sys import argv, stderr
from collections import namedtuple
from ad import Components, Context, tsm, t_const, t_mul

class Parameters(namedtuple('ParametersType', ['a', 'b', 'c'])):
    pass

def get_p():
    return Parameters(a = float(argv[8]),
                      b = float(argv[9]),
                      c = float(argv[10]))

def ode(x, y, z, p, k):
    return Components(x = - y[k] - z[k],
                      y = x[k] + p.a * y[k],
                      z = t_const(p.b, k) + t_mul(x, z, k) - p.c * z[k])

Context.places, order, h, steps = int(argv[1]), int(argv[2]), float(argv[3]), int(argv[4])  # controls
tsm(ode, Context.places, order, h, steps, float(argv[5]), float(argv[6]), float(argv[7]), get_p())
